New utility functions

SeungheonOh opened the pull request:
added `pisScriptAddress`, `pfindEffectInput`, and
`pfindEffectAddress`. These functions are commonly used by effects, like in #46 and #49.

chfanghr commented:
You may want to check this commit out: https://github.com/Liqwid-Labs/agora/commit/139f9f933a821a8d98b003c0a3a1f0e19eac9611

I made a bunch of change today, which might be more generic.

SeungheonOh commented:
Ahh how did I missed that, I'll update it

SeungheonOh commented:
@chfanghr committed the functions you mentioned, sorry it took long (because I was as school)...

chfanghr commented:
Don't worry; pls prioritize your school stuff if needed😄.

SeungheonOh commented:
@chfanghr I removed `findEffectInput` which was in fact doing same thing as `pfindTxInByTxOutRef`, fixed some comments, and reimplemented `findTxOutByTxOutRef`

chfanghr reviewed (approved):
LGTM!

emiflake commented on agora/Agora/Utils.hs:
```diff
@@ -390,3 +409,44 @@ psingletonValue = phoistAcyclic $
+scriptHashFromAddress :: Term s (PAddress :--> PMaybe PValidatorHash)
+scriptHashFromAddress = phoistAcyclic $
+  plam $ \addr -> P.do
+    cred <- pmatch $ pfromData $ pfield @"credential" # addr
```
Should probably inline `pmatch` here with a LambdaCase, then we also get rid of the `P.do`.

emiflake reviewed (changes requested):
Looks good! Just a few minor questions.
Ah, and fix the build please :sweat_smile:

chfanghr commented on agora/Agora/Utils.hs:
```diff
@@ -390,3 +409,44 @@ psingletonValue = phoistAcyclic $
+                  selfAddress #== address
+              )
+            # outputs
+    pmap @PList # plam pfromData #$ pconvertLists # filteredOutputs
```
Right, there's absolutely no need to return a `PList`.

chfanghr commented on agora/Agora/Utils.hs:
```diff
@@ -390,3 +409,44 @@ psingletonValue = phoistAcyclic $
+      _ -> pcon PNothing
+
+-- | Find all TxOuts sent to an Address
+findOutputsToAddress :: Term s (PTxInfo :--> PAddress :--> PBuiltinList PTxOut)
```
`PbuiltinList` can't contain non-builtin types. Should return `PbuiltinList (PAsData PtxOut)` here.

SeungheonOh replied:
Oh right 👍🏻

chfanghr commented on agora/Agora/Utils.hs:
```diff
@@ -390,3 +409,44 @@ psingletonValue = phoistAcyclic $
+                  selfAddress #== address
+              )
+            # outputs
+    pmap # plam pfromData #$ pconvertLists # filteredOutputs
```
This is unnecessary. Just return `filteredOutputs`.

chfanghr commented on agora/Agora/Utils.hs:
```diff
@@ -390,3 +409,44 @@ psingletonValue = phoistAcyclic $
+          pfilter
+            # plam
+              ( \(pfromData -> txOut) -> P.do
+                  selfAddress <- plet $ pfield @"address" # txOut
```
Oh I missed one. `pelt` here is redundant😢

SeungheonOh replied:
I already fixed that : )

SeungheonOh commented:
Finally, it looks good to be merged! 🎉

emiflake reviewed (approved):
Looks good!
